e2e_playwright/st_pydeck_chart_test.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+from playwright.sync_api import Page, expect
+
+from e2e_playwright.conftest import (
+    ImageCompareFunction,
+    wait_for_app_loaded,
+    wait_for_app_run,
+)
+
+
+# Firefox seems to be failing but can't reproduce locally and video produces an empty page for firefox
+@pytest.mark.skip_browser("firefox")
+def test_pydeck_chart_has_consistent_visuals(
+    app: Page, assert_snapshot: ImageCompareFunction
+):
+    pydeck_charts = app.get_by_test_id("stDeckGlJsonChart")
+    expect(pydeck_charts).to_have_count(4)
+
+    wait_for_app_run(app, 15000)
+    # The pydeck tests are a lot flakier than need be so increase the pixel threshold
+    assert_snapshot(
+        pydeck_charts.nth(0),
+        name="st_pydeck_chart-empty_light_theme",
+        pixel_threshold=1.0,
+    )
+
+    # The pydeck tests are a lot flakier than need be so increase the pixel threshold
+    assert_snapshot(
+        pydeck_charts.nth(1).locator("canvas").nth(0),
+        name="st_pydeck_chart-san_francisco_light_theme",
+        pixel_threshold=1.0,
+    )
+
+    # The pydeck tests are a lot flakier than need be so increase the pixel threshold
+    assert_snapshot(
+        pydeck_charts.nth(2).locator("canvas").nth(1),
+        name="st_pydeck_chart-continents_light_theme",
+        pixel_threshold=1.0,
+    )
+
+    # The pydeck tests are a lot flakier than need be so increase the pixel threshold
+    assert_snapshot(
+        pydeck_charts.nth(3).locator("canvas").nth(1),
+        name="st_pydeck_chart-geo_layers_light_theme",
+        pixel_threshold=1.0,
+    )
+
+
+# When using themed_app, it will navigate to a new page on each snapshot and refresh and thus unnecessary extra load time
+# Firefox seems to be failing but can't reproduce locally and video produces an empty page for firefox
```

Comment:
What do you mean with it will navigate to a new page on each snapshot and refresh and thus unnecessary extra load time? I don't think it would be a lot different from load time compared to having this split up into two methods with the same checks.

Reply:
I don't think it navigates to a new page for every individual screenshot. My expectation with `themed_app` is that it navigates to a new page for every test run (dark/light theme * browsers)

Reply:
>I don't think it would be a lot different from load time compared to having this split up into two methods with the same checks.
Hm when I was running it locally, it would call assert_snapshot on the light and then navigate to the dark version and take snapshots so I was assuming that's the way it works when running it on github actions too. However, technically, I was running `PWDEBUG=1 pytest e2e_playwright/st_pydeck_chart_test.py`? I would have to research the specific reason why this occurs (maybe because of fixtures but im not a fixture expert).

Reply:
Hm when I was running it locally, I misinterpretted the flow of execution. You're right.